src/Database/LSMTree/Internal/BloomFilterQuery2.hs

```diff
@@ -0,0 +1,441 @@
+{-# COMPLETE CandidateProbe #-}
+
+unpackCandidateProbe :: CandidateProbe -> (Int, Int, Int, Int)
+unpackCandidateProbe (MkCandidateProbe lo hi) =
+    ( fromIntegral ((lo `unsafeShiftR` 16) .&. 0xffff) -- run ix
+    , fromIntegral ( lo                    .&. 0xffff) -- key ix
+    , fromIntegral (lo `unsafeShiftR` 32)            -- hashno
+    , fromIntegral  hi
+    )
+{-# INLINE unpackCandidateProbe #-}
+
+packCandidateProbe :: Int -> Int -> Int -> Int -> CandidateProbe
+packCandidateProbe r k h p =
+    assert (r >= 0 && r <= 0xffff) $
+    assert (k >= 0 && k <= 0xffff) $
+    assert (h >= 0 && h <= 0xffff) $
+    MkCandidateProbe
+      (   fromIntegral r `unsafeShiftL` 16  -- run ix
+      .|. fromIntegral k                    -- key ix
+      .|. fromIntegral h `unsafeShiftL` 32) -- hashno
+      (fromIntegral p)
+{-# INLINE packCandidateProbe #-}
+
+instance Show CandidateProbe where
+  showsPrec _ (CandidateProbe r k h p) =
+      showString "CandidateProbe "
+    . showsPrec 11 r
+    . showChar ' '
+    . showsPrec 11 k
+    . showChar ' '
+    . showsPrec 11 h
+    . showChar ' '
+    . showsPrec 11 p
+
+
+
+bloomQueriesDefault ::
+     V.Vector (Bloom SerialisedKey)
+  -> V.Vector SerialisedKey
+  -> VP.Vector RunIxKeyIx
+bloomQueriesDefault filters keys | V.null filters || V.null keys
+                                 = VP.empty
+bloomQueriesDefault filters keys =
+  runST $ do
+    let !keyhashes = prepKeyHashes keys
+        !filters'  = P.vectorToStrictArray filters
+    candidateProbes <- P.newPrimArray 0x20
+--  P.setPrimArray 0 0x20 candidateProbes (MkCandidateProbe 0 0)
+
+    let i_r = 0
+    (i_w, nextCandidate) <-
+      prepInitialCandidateProbes
+        filters' keyhashes
+        candidateProbes
+        0 (RunIxKeyIx 0 0)
+
+    output <- bloomQueriesBody filters' keyhashes candidateProbes
+                               i_r i_w nextCandidate
+    return $! P.primArrayToPrimVector output
+
+prepKeyHashes :: V.Vector SerialisedKey
+              -> P.PrimArray (Bloom.CheapHashes SerialisedKey)
+prepKeyHashes keys =
+    P.generatePrimArray (V.length keys) $ \i ->
+      Bloom.makeCheapHashes (V.unsafeIndex keys i)
+
+prepInitialCandidateProbes
+  :: P.StrictArray (Bloom SerialisedKey)
+  -> P.PrimArray (Bloom.CheapHashes SerialisedKey)
+     -- ^ The pre-computed \"cheap hashes\" of the keys.
+  -> P.MutablePrimArray s CandidateProbe
+     -- ^ The array of candidates: of FilterIx, KeyIx and HashNo. This is
+     -- used as a fixed size rolling buffer.
+  -> Int
+     -- ^ The write index within the rolling buffer. This wraps around.
+  -> RunIxKeyIx
+     -- ^ The last candidate added to the rolling buffer. When this hits zero
+     -- then no more candidates are added, and we just drain the buffer.
+  -> ST s (Int,  RunIxKeyIx)
+prepInitialCandidateProbes
+    filters keyhashes
+    candidateProbes i_w
+    nextCandidate@(RunIxKeyIx rix kix)
+
+  | i_w < 0x1f && rix < P.sizeofStrictArray filters = do
```

Comment:
Yes it does. The condition is that once i_r == i_w then we declare the buffer empty and stop. So we can only use 31 of the 32 buffer entries. Hence we must prepare up to 31 at the start.

I agree this could be clearer. I was thinking of moving the buffer empty test to the start of `testCandidateProbe` rather than in `prepNextCandidateProbe`. Or to put it another way, it should be a `while (..) do {..}` loop, not a `do {..} while (..);` loop. This would make the invariants clearer, and should be much the same for code gen/performance.

Reply:
Added another comment to clarify.